File: gapic/src/main/com/google/gapid/glviewer/gl/Texture.java
/*
 * Copyright (C) 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.gapid.glviewer.gl;

import org.eclipse.swt.graphics.Color;
import org.lwjgl.opengl.GL11;
import org.lwjgl.opengl.GL13;
import org.lwjgl.opengl.GL33;

import java.nio.ByteBuffer;

/**
 * An OpenGL texture.
 */
public class Texture extends GlObject {
  private final int target;
  private final int handle;

  Texture(Renderer owner, int target) {
    super(owner);
    this.target = target;
    this.handle = GL11.glGenTextures();
    owner.register(this);
  }

  public Texture setMinMagFilter(int min, int mag) {
    bind();
    GL11.glTexParameteri(target, GL11.GL_TEXTURE_MIN_FILTER, min);
    GL11.glTexParameteri(target, GL11.GL_TEXTURE_MAG_FILTER, mag);
    return this;
  }

  public Texture setBorderColor(Color color) {
    bind();
    GL11.glTexParameterfv(target, GL11.GL_TEXTURE_BORDER_COLOR, new float[]{
        color.getRed() / 255.f,
        color.getGreen() / 255.f,
        color.getBlue() / 255.f,
        color.getAlpha() / 255.f,
    });
    return this;
  }

  public Texture setWrapMode(int wrapS, int wrapT) {
    bind();
    GL11.glTexParameteri(target, GL11.GL_TEXTURE_WRAP_S, wrapS);
    GL11.glTexParameteri(target, GL11.GL_TEXTURE_WRAP_T, wrapT);
    return this;
  }

  public Texture setSwizzle(int r, int g, int b, int a) {
    bind();
    GL11.glTexParameteriv(target, GL33.GL_TEXTURE_SWIZZLE_RGBA, new int[] { r, g, b, a });
    return this;
  }

  public Texture loadData(
      int width, int height, int internalFormat, int format, int type, ByteBuffer data) {
    bind();
    GL11.glPixelStorei(GL11.GL_UNPACK_ALIGNMENT, 1);
    GL11.glTexImage2D(target, 0, internalFormat, width, height, 0, format, type, data);
    return this;
  }

  static void activate(int unit) {
    GL13.glActiveTexture(GL13.GL_TEXTURE0 + unit);
  }

  void bind() {
    GL11.glBindTexture(target, handle);
  }

  @Override
  protected void release() {
    GL11.glDeleteTextures(handle);
  }
}

File: gapis/gfxapi/texture.go
// Copyright (C) 2017 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package gfxapi

import (
	"context"

	"github.com/google/gapid/core/image"
)

func (l *CubemapLevel) faces() [6]*image.Info2D {
	return [6]*image.Info2D{
		l.NegativeX,
		l.PositiveX,
		l.NegativeY,
		l.PositiveY,
		l.NegativeZ,
		l.PositiveZ,
	}
}

func (l *CubemapLevel) setFaces(faces [6]*image.Info2D) {
	l.NegativeX,
		l.PositiveX,
		l.NegativeY,
		l.PositiveY,
		l.NegativeZ,
		l.PositiveZ = faces[0], faces[1], faces[2], faces[3], faces[4], faces[5]
}

type imageMatcher struct {
	best          *image.Info2D
	score         uint32
	width, height uint32
}

func (m *imageMatcher) consider(i *image.Info2D) {
	if m.best == nil {
		m.score = 0xffffffff
	}
	dw, dh := i.Width-m.width, i.Height-m.height
	score := dw*dw + dh*dh
	if m.score > score {
		m.score = score
		m.best = i
	}
}

// Thumbnail returns the image that most closely matches the desired size.
func (t *Texture2D) Thumbnail(ctx context.Context, w, h uint32) (*image.Info2D, error) {
	m := imageMatcher{width: w, height: h}
	for _, l := range t.Levels {
		m.consider(l)
	}

	return m.best, nil
}

// ConvertTo returns this Texture2D with each mip-level converted to the requested format.
func (t *Texture2D) ConvertTo(ctx context.Context, f *image.Format) (interface{}, error) {
	out := &Texture2D{
		Levels: make([]*image.Info2D, len(t.Levels)),
	}
	for i, m := range t.Levels {
		if obj, err := m.ConvertTo(ctx, f); err == nil {
			out.Levels[i] = obj
		} else {
			return nil, err
		}
	}
	return out, nil
}

// Thumbnail returns the image that most closely matches the desired size.
func (t *Cubemap) Thumbnail(ctx context.Context, w, h uint32) (*image.Info2D, error) {
	m := imageMatcher{width: w, height: h}

	for _, l := range t.Levels {
		m.consider(l.NegativeX)
		m.consider(l.PositiveX)
		m.consider(l.NegativeY)
		m.consider(l.PositiveY)
		m.consider(l.NegativeZ)
		m.consider(l.PositiveZ)
	}

	return m.best, nil
}

// ConvertTo returns this Cubemap with each mip-level face converted to the requested format.
func (t *Cubemap) ConvertTo(ctx context.Context, f *image.Format) (interface{}, error) {
	out := &Cubemap{
		Levels: make([]*CubemapLevel, len(t.Levels)),
	}
	for i, m := range t.Levels {
		out.Levels[i] = &CubemapLevel{}
		dst, src := out.Levels[i].faces(), m.faces()
		for j := range src {
			if obj, err := src[j].ConvertTo(ctx, f); err == nil {
				dst[j] = obj
			} else {
				return nil, err
			}
		}
		out.Levels[i].setFaces(dst)
	}
	return out, nil
}
